extra/innochecksum.cc

```diff
@@ -873,6 +873,17 @@ parse_page(
 		break;
 	}
 	case FIL_PAGE_UNDO_LOG:
+		{
+			page_no = mach_read_from_4(page + FIL_PAGE_OFFSET);
+			const byte *des= xdes + XDES_ARR_OFFSET +
+				XDES_SIZE *
+				((page_no & (physical_page_size - 1))
+				 / FSP_EXTENT_SIZE);
+			if (mach_read_from_4(des) != XDES_FSEG &&
+		            xdes_is_free(des, page_no % FSP_EXTENT_SIZE)) {
+				break;
+			}
+		}
```

Comment:
I don’t think that it is a good idea to make this a special case for a particular page type. In older InnoDB data files, the `FIL_PAGE_TYPE` could be garbage. The only thing that held was that if the page is an index page, then `FIL_PAGE_TYPE` would be written as `FIL_PAGE_INDEX`. The opposite was not necessarily true: any page could carry a `FIL_PAGE_TYPE` of `FIL_PAGE_INDEX`.

It would seem better to implement this logic for any page type, and enabled by a special command line option of `innochecksum`, say, `--skip-freed-pages`. In that way, this feature would also be opt-in.